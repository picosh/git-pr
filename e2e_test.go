package git

import (
	"log/slog"
	"os"
	"testing"
	"time"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/picosh/git-pr/fixtures"
	"github.com/picosh/git-pr/util"
)

func TestE2E(t *testing.T) {
	testSingleTenantE2E(t)
	testMultiTenantE2E(t)
}

func testSingleTenantE2E(t *testing.T) {
	t.Log("single tenant end-to-end tests")
	dataDir := util.CreateTmpDir()
	defer func() {
		os.RemoveAll(dataDir)
	}()
	suite := setupTest(dataDir, cfgSingleTenantTmpl)
	done := make(chan error)
	go GitSshServer(suite.cfg, done)
	// Hack to wait for startup
	time.Sleep(time.Millisecond * 100)

	t.Log("User cannot create repo")
	_, err := suite.userKey.Cmd(suite.patch, "pr create test")
	if err == nil {
		t.Fatal("user should not be able to create a PR")
	}
	suite.adminKey.MustCmd(suite.patch, "pr create test")

	t.Log("User should be able to create a patch")
	suite.userKey.MustCmd(suite.patch, "pr create test")

	t.Log("Snapshot test ls command")
	actual, err := suite.userKey.Cmd(nil, "pr ls")
	bail(err)
	snaps.MatchSnapshot(t, actual)

	done <- nil
}

func testMultiTenantE2E(t *testing.T) {
	t.Log("multi tenant end-to-end tests")
	dataDir := util.CreateTmpDir()
	defer func() {
		os.RemoveAll(dataDir)
	}()
	suite := setupTest(dataDir, cfgMultiTenantTmpl)
	done := make(chan error)
	go GitSshServer(suite.cfg, done)

	time.Sleep(time.Millisecond * 100)

	t.Log("Admin should be able to create a repo")
	suite.adminKey.MustCmd(nil, "repo create test")

	t.Log("Accepted pr")
	suite.userKey.MustCmd(suite.patch, "pr create admin/test")
	suite.userKey.MustCmd(nil, "pr edit 1 Accepted patch")
	_, err := suite.userKey.Cmd(nil, "pr accept 1")
	if err == nil {
		t.Fatal("contrib should not be able to accept their own PR")
	}
	suite.adminKey.MustCmd(nil, "pr accept 1")

	t.Log("Closed pr (admin)")
	suite.userKey.MustCmd(suite.patch, "pr create test")
	suite.userKey.MustCmd(nil, "pr edit 2 Closed patch (admin)")
	suite.adminKey.MustCmd(nil, "pr close 2")

	t.Log("Closed pr (contributor)")
	suite.userKey.MustCmd(suite.patch, "pr create test")
	suite.userKey.MustCmd(nil, "pr edit 3 Closed patch (contributor)")
	suite.userKey.MustCmd(nil, "pr close 3")

	t.Log("Reviewed pr")
	suite.userKey.MustCmd(suite.patch, "pr create test")
	suite.userKey.MustCmd(nil, "pr edit 4 Reviewed patch")
	suite.adminKey.MustCmd(suite.otherPatch, "pr add --review 4")

	t.Log("Accepted pr with review")
	suite.userKey.MustCmd(suite.patch, "pr create test")
	suite.userKey.MustCmd(nil, "pr edit 5 Accepted patch with review")
	suite.adminKey.MustCmd(suite.otherPatch, "pr add --accept 5")

	t.Log("Closed pr with review")
	suite.userKey.MustCmd(suite.patch, "pr create test")
	suite.userKey.MustCmd(nil, "pr edit 6 Closed patch with review")
	suite.adminKey.MustCmd(suite.otherPatch, "pr add --close 6")

	t.Log("Create pr with user repo and user can accept")
	suite.userKey.MustCmd(nil, "repo create ai")
	suite.adminKey.MustCmd(suite.patch, "pr create contributor/ai")
	suite.userKey.MustCmd(suite.otherPatch, "pr accept 7")

	t.Log("Create pr with user repo and admin can accept")
	suite.adminKey.MustCmd(nil, "repo create ai")
	suite.userKey.MustCmd(suite.patch, "pr create admin/ai")
	suite.adminKey.MustCmd(suite.otherPatch, "pr add --accept 8")

	t.Log("Create pr with default `bin` repo")
	actual, err := suite.userKey.Cmd(suite.patch, "pr create")
	bail(err)
	snaps.MatchSnapshot(t, actual)

	t.Log("Snapshot test ls command")
	actual, err = suite.userKey.Cmd(nil, "pr ls")
	bail(err)
	snaps.MatchSnapshot(t, actual)

	t.Log("Snapshot test logs command")
	actual, err = suite.userKey.Cmd(nil, "logs --repo admin/ai")
	bail(err)
	snaps.MatchSnapshot(t, actual)

	done <- nil
}

type TestSuite struct {
	cfg        *GitCfg
	userKey    util.UserSSH
	adminKey   util.UserSSH
	patch      []byte
	otherPatch []byte
}

func setupTest(dataDir string, cfgTmpl string) TestSuite {
	opts := &slog.HandlerOptions{
		AddSource: true,
	}
	logger := slog.New(
		slog.NewTextHandler(os.Stdout, opts),
	)

	adminKey, userKey := util.GenerateKeys()
	cfgPath := util.CreateCfgFile(dataDir, cfgTmpl, adminKey)
	LoadConfigFile(cfgPath, logger)
	cfg := NewGitCfg(logger)

	// so outputs dont show dates
	cfg.TimeFormat = ""

	patch, err := fixtures.Fixtures.ReadFile("single.patch")
	if err != nil {
		panic(err)
	}
	otherPatch, err := fixtures.Fixtures.ReadFile("with-cover.patch")
	if err != nil {
		panic(err)
	}

	return TestSuite{cfg, userKey, adminKey, patch, otherPatch}
}

var cfgSingleTenantTmpl = `
url = "localhost"
data_dir = %q
admins = [%q]
time_format = "01/02/2006 15:04:05 07:00"
create_repo = "admin"`

var cfgMultiTenantTmpl = `
url = "localhost"
data_dir = %q
admins = [%q]
time_format = "01/02/2006 15:04:05 07:00"
create_repo = "user"`
