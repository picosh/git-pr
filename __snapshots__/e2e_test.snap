
[TestE2E - 1]
ID RepoID Name                    Status Patchsets User        Date
2  test   feat: lets build an rnn [open] 1         contributor 
1  test   feat: lets build an rnn [open] 1         admin       

---

[TestE2E - 2]
PR submitted! Use the ID for interacting with this PR.
Info
====
URL: https://localhost/prs/9
Repo: contributor/bin

ID Name                    Status Date
9  feat: lets build an rnn [open] 

Patchsets
====
ID    Type User        Date
ps-13      contributor 

Patches from latest patchset
====
Idx Title                   Commit  Author                   Date
0   feat: lets build an rnn 5945657 Eric Bower <me@erock.io> 

---

[TestE2E - 3]
ID RepoID           Name                       Status     Patchsets User        Date
9  contributor/bin  feat: lets build an rnn    [open]     1         contributor 
8  admin/ai         feat: lets build an rnn    [accepted] 2         contributor 
7  contributor/ai   feat: lets build an rnn    [accepted] 1         admin       
6  contributor/test Closed patch with review   [closed]   2         contributor 
5  contributor/test Accepted patch with review [accepted] 2         contributor 
4  contributor/test Reviewed patch             [reviewed] 2         contributor 
3  contributor/test Closed patch (contributor) [closed]   1         contributor 
2  contributor/test Closed patch (admin)       [closed]   1         contributor 
1  admin/test       Accepted patch             [accepted] 1         contributor 

---

[TestE2E - 4]
RepoID   PrID PatchsetID Event             Created Data
admin/ai 8    ps-11      pr_created                
admin/ai 8    ps-12      pr_patchset_added         
admin/ai 8               pr_status_changed         {"status":"accepted"}

---
